scipy/optimize/jonkervolgenant/jonkervolgenant.cpp

```diff
@@ -0,0 +1,194 @@
+/*
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+   notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above
+   copyright notice, this list of conditions and the following
+   disclaimer in the documentation and/or other materials provided
+   with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+   contributors may be used to endorse or promote products derived
+   from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+This code implements the Jonker-Volgenant algorithm for the rectangular
+assignment problem:
+
+    R. Jonker and A. Volgenant. A Shortest Augmenting Path Algorithm for
+    Dense and Sparse Linear Assignment Problems. *Computing*, 38:325-340
+    December 1987.
+
+The code is based on the pseudocode described in pages
+1685-1686 of:
+
+    DF Crouse. On implementing 2D rectangular assignment algorithms.
+    IEEE Transactions on Aerospace and Electronic Systems
+    52(4):1679-1696, August 2016
+    doi: 10.1109/TAES.2016.140952
+
+Author: PM Larsen
+*/
+
+#include <algorithm>
+#include <cmath>
+#include <vector>
+
+static int
+augmenting_path(int nc, std::vector<double>& cost, std::vector<double>& u,
+                std::vector<double>& v, std::vector<int>& path,
+                std::vector<int>& row4col,
+                std::vector<double>& shortestPathCosts, int i,
+                std::vector<bool>& SR, std::vector<bool>& SC, double* p_minVal)
+{
+    double minVal = 0;
+
+    int num_remaining = nc;
+    std::vector<int> remaining(nc);
```

Comment:
It could be useful to have a comment on what `remaining` does, simply because that one isn't covered in detail in the IEEE review while everything else is.

I assume it's included mainly to improve performance? And it certainly looks like it does. If I compare to the naive implementation of just using the code snippet below on a 5000x5000 example, I get the following results (10 runs each):
```
Approach from IEEE review: Mean: 4.2826s, std: 0.295584s
Using "remaining" array:   Mean: 3.4091s, std: 0.35081s
```

```cpp
int index = -1;
double lowest = INFINITY;
for (int j = 0; j < nc; j++)
{
	if (SC[j]) continue;
	double r = minVal + cost[i * nc + j] - u[i] - v[j];
	if (r < shortestPathCosts[j]) {
		path[j] = i;
		shortestPathCosts[j] = r;
	}
	if (shortestPathCosts[j] < lowest || (shortestPathCosts[j] == lowest && row4col[j] == -1))
	{
		index = j;
		lowest = shortestPathCosts[j];
	}

}
int j = index;
minVal = lowest;
```

Reply:
Yes, it is for performance reasons.  I have added a comment to explain this.